libs/prelude/Prelude/Types.idr

```diff
@@ -71,16 +84,11 @@ Num Nat where
 
 public export
 Eq Nat where
-  Z == Z = True
-  S j == S k = j == k
-  _ == _ = False
+  (==) = equal
```

Comment:
As mentioned in #1182, can't we use a `transform` rule instead?

%transform "nateq" (==) {a = Nat} m n = natToInteger n == natToInteger m

Reply:
I was testing this out, `%transform` doesn't seem to work all the time. 

Here is an example. As you can see, `fee5` calls `fee` directly whereas `fee3` calls `fee2` which in turn calls `fee`. For the `fee5` call, the result is transformed at runtime, for the `fee3` call, it is not.

```
interface Foo x where
  fee : x -> String
  
data Bar = Bee | Boot

Foo Bar where
  fee Bee = "Bee"
  fee Boot = "Boot"

%transform "feechange" fee {x=Bar} x = "No matter"

fee2 : Foo x => x -> String
fee2 v = fee v

fee3 : Bar -> String
fee3 v = fee2 v

fee5 : Bar -> String
fee5 v = fee v

test : IO ()
test = do 
         --using :exec test 
         --at runtime, prints:
         -- fee3 : Bee
         -- fee5 : No matter

         putStrLn $ "fee3 : " ++ fee3 Bee
         putStrLn $ "fee5 : " ++ fee5 Bee
```

Reply:
If `%transform` is unreliable shouldn't we try and fix that rather than introduce more compiler magic that the end user can't extend in their own code?

Reply:
I was just thinking that it seemed that the hacks general purpose was to speed up Nat computation, which it wasn't doing. So, fixing the hack wouldn't introduce significant more dirtiness into the code that was already there, and was easy to do.

IMO, getting something working quickly is better since it attracts more users, but I don't know if the idris developers are looking for more users right now, so maybe this isn't a goal of theirs. 

However, the best would probably be to rip out the hack completely and use %transform for everything, including minus,+, and * for which the hack already works.